contracts/contracts/orand/OrandProviderV1.sol

```diff
@@ -35,13 +36,27 @@ contract OrandProviderV1 is OrandStorage, OrandManagement, OrandSignatureVerifie
   //=======================[  External  ]====================
   // Publish new epoch
   function publish(bytes memory proof, EpochProof memory newEpoch) external onlyReadyForPenalty returns (bool) {
-    (bool verified, address verifierAddress, uint256 y) = _vefifyProof(proof);
-    require(verified, 'OP1: Invalid proof');
-    require(y == newEpoch.y, 'OP1: Invalid ERCVRF output');
-    require(_addEpoch(verifierAddress, newEpoch), 'OP1: Unable to add new epoch');
-    require(IOrandConsumerV1(verifierAddress).consumeRandomness(newEpoch.y), 'OP1: Unable to send the randomness');
-    // Increase nonce of receiver to prevent replay attack
-    _increaseNonce(verifierAddress);
+    (bool verified, address receiverAddress, uint256 y) = _vefifyProof(proof);
+    // Verifier is false, signature proof is incorrect
+    if (!verified) {
+      revert InvalidProof(proof);
+    }
+    // Linked y is different from submitted value
+    if (y != newEpoch.y) {
+      revert InvalidECVRFOutput(y, newEpoch.y);
+    }
+    // Unable to add epoch to storage
+    if (!_addEpoch(receiverAddress, newEpoch)) {
+      revert UnableToAddNewEpoch(receiverAddress, newEpoch);
+    }
+    // Unable to forward randomness to receiver contract
+    if (!IOrandConsumerV1(receiverAddress).consumeRandomness(newEpoch.y)) {
+      revert UnableToForwardRandomness(receiverAddress, y);
+    }
+    // Increasing nonce of receiver to prevent replay attack
+    if (!_increaseNonce(receiverAddress)) {
```

Comment:
I think something wrong in this logic. `_increaseNonce` always return true.

Reply:
Just want to handle unexpected case, it looks unnecessary but it made me feel safe.

```solidity
require(_increaseNonce(receiverAddress), "It suppose to work");
```